shared/src/commonMain/kotlin/io/customer/shared/di/DIGraph.kt

```diff
@@ -0,0 +1,101 @@
+    /**
+     * Map of overridden dependencies, helpful in mocking for test functions.
+     */
+    val overrides: MutableMap<String, Any> = mutableMapOf()
+
+    /**
+     * Call when you want to override a dependency with a mock in test functions.
+     * `di.overrideDependency(CustomerIOInstance::class.java, customerIOMock)`
+     *
+     * See `CustomerIOTest` (or any other class that is top-level of the module) for an example on use.
+     */
+    fun <DEP : Any> overrideDependency(dependency: KClass<DEP>, value: DEP) {
+        overrides[dependency.getClassNameOrThrow(value = value)] = value as Any
+    }
+
+    /**
+     * Call in Digraph property getters to get allow mock overriding before trying to get actual instance.
+     * We strongly recommend to get instances using methods exposed by Digraph for the purpose, so
+     * it can take responsibility of providing overridden instances where needed.
+     * However, if needed, actual instance should be accessed using the following:
+     * ```
+     * val fileStorage: FileStorage
+     *   get() = override() ?: FileStorage(siteId, context)
+     * ```
+     */
+    inline fun <reified DEP> override(): DEP? = overrides[DEP::class.simpleName] as? DEP
+
+    /**
+     * We prefer to have all of the SDK's singleton instances held in the dependency injection graph.
+     * This makes it easier for automated tests to be able to delete all singletons between each
+     * test function and prevent test flakiness.
+     */
+    val singletons: MutableMap<String, Any> = mutableMapOf()
+
+    /**
+     * In the graph, if you have any dependency that should be a singleton:
+     * ```
+     * val queue: Queue
+     *   get() = getSingletonInstance { QueueImpl(...) }
+     * ```
+     * Note: Don't forget to include the interface in the property declaration:
+     * ```
+     * val foo: InterfaceOfFoo
+     *          ^^^^^^^^^^^^^^
+     * ```
+     * Or you may not be able to successfully mock in test functions.
+     */
+    inline fun <reified INST : Any> getSingletonInstance(crossinline newInstanceCreator: () -> INST): INST {
+        val singletonKey = INST::class.getClassNameOrThrow()
+        val newInstance = { override() ?: newInstanceCreator() }
+        return singletons[singletonKey] as? INST ?: newInstance().also { instance ->
+            singletons[singletonKey] = instance
+        }
+    }
+
+    /**
+     * In the graph, if you have any dependency that should create new instance every time:
+     * ```
+     * val queue: Queue
+     *   get() = getNewInstance { QueueImpl(...) }
+     * ```
+     * Note: Don't forget to include the interface in the property declaration:
+     * ```
+     * val foo: InterfaceOfFoo
+     *          ^^^^^^^^^^^^^^
+     * ```
+     * Or you may not be able to successfully mock in test functions.
+     */
+    inline fun <reified INST : Any> getNewInstance(newInstanceCreator: () -> INST): INST {
+        return override() ?: newInstanceCreator()
+    }
```

Comment:
What is the reason for adding this? 

Does this work in Kotlin? 

```kt
class DiGraph {
  val queue: Queue
    get() = QueueImpl(...)
}

// each time that I call this code:
diGraphInstance.queue
// I get a new instance of `Queue`? 
```

Reply:
I'm not sure if I understand your question correctly but still, here are my thoughts

> // I get a new instance of `Queue`? 

Yes, if you use `get() =` in your property, it will create new every time.

> What is the reason for adding this?

Mainly to eliminate repetitive calling of `override()` at every creation. This method helps keeping it simple.

Reply:
> Mainly to eliminate repetitive calling of override() at every creation. This method helps keeping it simple.

Ah, I do like that idea. 

I would, however, change the name of the function then to be more descriptive. 

Idea: `overrideOrNewInstance()`

Reply:
I avoided the name describing this behavior by assuming the following

- Our DIGraph works with assumption of overriding taking preference for all methods of creating objects (both singleton and new instance)
- We mainly use this for tests, so reflecting the behavior in name doesn't look like a very good idea to me

Do you still think we should rename this? And if we do, we'll have to rename both like `overrideOrNewInstance` and `overrideOrSingletonInstance`